src/audio/crossover/crossover.c

```diff
@@ -123,56 +124,61 @@ static int crossover_get_stream_index(struct sof_crossover_config *config,
  * \return number of sinks assigned. This number should be equal to
  *	   config->num_sinks if no errors were found.
  */
-static int crossover_assign_sinks(struct comp_dev *dev,
-				  struct sof_crossover_config *config,
-				  struct comp_buffer **sinks)
+static int crossover_assign_sinks(struct processing_module *mod,
+				  struct comp_buffer **sinks,
+				  bool *enabled)
 {
+	struct comp_data *cd = module_get_private_data(mod);
+	struct sof_crossover_config *config = cd->config;
+	struct comp_dev *dev = mod->dev;
 	struct comp_buffer *sink;
 	struct comp_buffer __sparse_cache *sink_c;
 	struct list_item *sink_list;
 	int num_sinks = 0;
 	int i;
+	int j = 0;
 
-	/* Align sink streams with their respective configurations */
 	list_for_item(sink_list, &dev->bsink_list) {
 		unsigned int pipeline_id, state;
 
 		sink = container_of(sink_list, struct comp_buffer, source_list);
 		sink_c = buffer_acquire(sink);
-
 		pipeline_id = sink_c->pipeline_id;
 		state = sink_c->sink->state;
 		buffer_release(sink_c);
-
-		if (state != dev->state)
+		if (state != dev->state) {
+			j++;
 			continue;
+		}
 
 		/* If no config is set, then assign the sinks in order */
 		if (!config) {
 			sinks[num_sinks++] = sink;
+			enabled[j++] = true;
 			continue;
 		}
 
-		i = crossover_get_stream_index(config, pipeline_id);
+		i = crossover_get_stream_index(mod, config, pipeline_id);
 
 		/* If this sink buffer is not assigned
 		 * in the configuration.
 		 */
 		if (i < 0) {
 			comp_err(dev,
-				 "crossover_acquire_sinks(), could not find sink %d in config",
+				 "crossover_assign_sinks(), could not find sink %d in config",
 				 pipeline_id);
 			break;
 		}
 
 		if (sinks[i]) {
 			comp_err(dev,
-				 "crossover_acquire_sinks(), multiple sinks from pipeline %d are assigned",
+				 "crossover_assign_sinks(), multiple sinks from pipeline %d are assigned",
 				 pipeline_id);
 			break;
 		}
 
 		sinks[i] = sink;
+		enabled[j++] = true;
```

Comment:
these changes seem more like a bugfix rather than a IPC4 change, can you split into a different commit?

Reply:
I can't see the bug, is there one? The enabled vector is for correctly updating mod->output_buffers[i].size. I'm not sure why the multiple buffers are provided by module API like they are now with possibly different state than component. It would be nice if the slow list search could be avoided totally and only the buffers vector from module api could be used.